### PR TITLE
[SPARK-38932][SQL] Datasource v2 support report distinct keys

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

- Add a new mix in interface `SupportsReportDistinctKeys` for datasource v2
- Add a new method `reportDistinctKeysSet` in `LeafNode`
- Override `reportDistinctKeysSet` in datasource v2 relation
- Propagate `reportDistinctKeysSet` at `DistinctKeysVisitor`

### Why are the changes needed?

Datasource v2 can be used to connect to some databases who support [unique key](https://en.wikipedia.org/wiki/Unique_key).

Spark catalyst optimizer support do further optimization through distinct keys. So it can improve the performance if the Scan reports its distinct keys to Spark.

We already have several optimizer rules for distinct keys, for example:
- https://github.com/apache/spark/pull/35779
- https://github.com/apache/spark/pull/36117
- https://github.com/apache/spark/pull/36530

We also have some prs which is in progress related distinct keys,  for example:
- https://github.com/apache/spark/pull/37267
- https://github.com/apache/spark/pull/36180

### Does this PR introduce _any_ user-facing change?

yes, a new interface added for developer

### How was this patch tested?

add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @sigmod @wangyum

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/SupportsReportUniqueKeys.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+package org.apache.spark.sql.connector.read;
+
+import java.util.Set;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.expressions.NamedReference;
+
+/**
+ * A mix in interface for {@link Scan}. Data sources can implement this interface to
+ * report unique keys set to Spark.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/SupportsReportUniqueKeys.java:27]: Can we also explain that what Spark SQL does with the unique keys if we implement this?

Comment (ulysses-you) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/SupportsReportUniqueKeys.java:27]: sure, will add some details about how Spark does with the unique keys

**Comment by ulysses-you** [comment]
cc @cloud-fan @huaxingao if you have time to take a look, thank you

**Pull request closed without merging**